### PR TITLE
New password reset flow

### PR DESCRIPTION
**Pull request by weeman1337**
Implements the new password reset flow on web.

Closes https://github.com/vector-im/element-web/issues/23131

PSF-1543

https://www.figma.com/file/C8zAfownE0T5pvyaJuwQK0/[Log-in]-Forgot-password

Before 

https://user-images.githubusercontent.com/6216686/202712228-8d1327ca-35a2-4400-8ba4-fbaedcb3b2fb.mp4

After

https://user-images.githubusercontent.com/6216686/202711626-b222c152-785f-4d5d-b58f-0a1535a77da5.mp4

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * New password reset flow ([\#9581](https://github.com/matrix-org/matrix-react-sdk/pull/9581)). Fixes vector-im/element-web#23131.

### CODE REVIEW

File: src/PasswordReset.ts [hunk 1]
```diff
@@ -61,7 +64,27 @@ export default class PasswordReset {
     ): Promise<IRequestTokenResponse> {
         this.password = newPassword;
         this.logoutDevices = logoutDevices;
-        return this.client.requestPasswordEmailToken(emailAddress, this.clientSecret, 1).then((res) => {
+        this.sendAttempt++;
+        return this.client.requestPasswordEmailToken(emailAddress, this.clientSecret, this.sendAttempt).then((res) => {
+            this.sessionId = res.sid;
+            return res;
+        }, function(err) {
```
Comment (turt2live) [inline src/PasswordReset.ts:71]: we should be using arrow functions, and `.catch()` in new/modified code.

File: src/PasswordReset.ts [hunk 2]
```diff
@@ -98,7 +144,7 @@ export default class PasswordReset {
                 threepid_creds: creds,
                 threepidCreds: creds,
             }, this.password, this.logoutDevices);
-        } catch (err) {
+        } catch (err: any) {
```
Comment (turt2live) [inline src/PasswordReset.ts:147]: Proper type please

File: src/components/structures/auth/ForgotPassword.tsx [hunk 3]
```diff
@@ -17,130 +17,98 @@ limitations under the License.
...
-
-export default class ForgotPassword extends React.Component<IProps, IState> {
+export default class ForgotPassword extends React.Component<Props, State> {
     private reset: PasswordReset;
+    private fieldPassword: Field | null = null;
+    private fieldPasswordConfirm: Field | null = null;
 
-    state: IState = {
-        phase: Phase.Forgot,
+    state: State = {
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:83]: we should be setting this in the constructor, if we're in the area doing refactorings.

File: src/components/structures/auth/ForgotPassword.tsx [hunk 4]
```diff
@@ -154,94 +122,77 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
...
+
+    private sendVerificationMail = async (): Promise<boolean> => {
         try {
-            await this.handleHttpRequest(this.reset.checkEmailLinkClicked());
-            this.setState({ phase: Phase.Done });
-        } catch (err) {
-            this.showErrorDialog(err.message);
+            await this.reset.requestResetToken(this.state.email);
+            return true;
+        } catch (err: any) {
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:140]: proper types please (applies throughout)

File: src/components/structures/auth/forgot-password/EnterEmail.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    };
+
+    return <>
+        <EMailIcon className="mx_AuthBody_icon" />
+        <h1>{ _t("Enter your email to reset password") }</h1>
+        <p className="mx_AuthBody_text">
+            {
+                _t(
+                    "<b>%(homeserver)s</b> will send you a verification link to let you reset your password.",
+                    { homeserver: "matrix.org" },
```
Comment (turt2live) [inline src/components/structures/auth/forgot-password/EnterEmail.tsx:67]: This shouldn't be hardcoded?

**Review by turt2live — CHANGES_REQUESTED** [review]
This is a first pass review - the change is complex and appears to mix refactoring with functionality (please avoid doing this).

Screenshots would be very appreciated for both historical context and to aid review.

**Comment by weeman1337** [comment]
> This is a first pass review - the change is complex and appears to mix refactoring with functionality (please avoid doing this).

Only thing here that does not belong to the new reset flow is fixing the field tooltip position. I will pull this out. Otherwise it is difficult to deliver parts of the new reset flow. Anyway I will try to split things even more in the future.

> Screenshots would be very appreciated for both historical context and to aid review.

I am going to add screen recordings of both.

File: src/components/structures/auth/ForgotPassword.tsx (hunk 3 above)
Comment (weeman1337) [reply src/components/structures/auth/ForgotPassword.tsx:83]: Done

File: src/PasswordReset.ts (hunk 1 above)
Comment (weeman1337) [reply src/PasswordReset.ts:71]: Done

File: src/components/structures/auth/forgot-password/EnterEmail.tsx (hunk 5 above)
Comment (weeman1337) [reply src/components/structures/auth/forgot-password/EnterEmail.tsx]: Oh, yes sure. Matrix.org is THE homeserver :upside_down_face: 

Good catch :+1: Fixing this and update the tests…

File: src/PasswordReset.ts (hunk 2 above)
Comment (weeman1337) [reply src/PasswordReset.ts:153]: Marking this as resolved to continue discussion here https://github.com/matrix-org/matrix-react-sdk/pull/9581#discussion_r1025756523

File: src/components/structures/auth/ForgotPassword.tsx (hunk 4 above)
Comment (turt2live) [reply src/components/structures/auth/ForgotPassword.tsx:176]: huh, that's both interesting and annoying. We might consider using `unknown` instead, but that's a conversation for another PR or at least another time :P

File: src/components/structures/auth/ForgotPassword.tsx [hunk 6]
```diff
@@ -136,7 +134,14 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
                 serverIsAlive: true,
             });
         } catch (e) {
-            this.setState(AutoDiscoveryUtils.authComponentStateForError(e, "forgot_password") as IState);
+            const {
+                serverIsAlive,
+                serverDeadError,
+            } = AutoDiscoveryUtils.authComponentStateForError(e, "forgot_password");
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:140]: the TS strict check on this line might want `e` to be defined as `: any`, I guess.

File: src/components/structures/auth/forgot-password/EnterEmail.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            onSubmitForm(event);
+            return;
+        }
+
+        emailFieldRef.current?.focus();
+        emailFieldRef.current?.validate({ allowEmpty: false, focused: true });
+    };
+
+    return <>
+        <EMailIcon className="mx_AuthBody_icon" />
```
Comment (turt2live) [inline src/components/structures/auth/forgot-password/EnterEmail.tsx:63]: We count `Email` as a single word - `M` should be lowercase in this class name (to be at least consistent with `EmailField` and etc)

File: src/components/structures/auth/forgot-password/VerifyEmailModal.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+}) => {
+    const { toggle: toggleTooltipVisible, value: tooltipVisible } = useTimeoutToggle(false, 2500);
+
+    const onResendClickFn = async (): Promise<void> => {
+        await onResendClick();
+        toggleTooltipVisible();
+    };
+
+    return <>
+        <EMailPromptIcon className="mx_AuthBody_emailPromptIcon" />
```
Comment (turt2live) [inline src/components/structures/auth/forgot-password/VerifyEmailModal.tsx:46]: `Email`

**Review by turt2live — COMMENTED** [review]
otherwise lgtm - thanks!

**Review by turt2live — APPROVED** [review]

**Pull request merged**